discoveryservice/store.go

```diff
@@ -0,0 +1,367 @@
+		PresentationExpiration: presentation.JWT().Expiration().Unix(),
+	}
+
+	for _, currCred := range presentation.VerifiableCredential {
+		var credentialType *string
+		for _, currType := range currCred.Type {
+			if currType.String() != "VerifiableCredential" {
+				credentialType = new(string)
+				*credentialType = currType.String()
+				break
+			}
+		}
+		if len(currCred.CredentialSubject) != 1 {
+			return nil, errors.New("credential must contain exactly one subject")
+		}
+
+		newCredential := credentialRecord{
+			ID:                  uuid.NewString(),
+			PresentationID:      newPresentation.ID,
+			CredentialID:        currCred.ID.String(),
+			CredentialIssuer:    currCred.Issuer.String(),
+			CredentialSubjectID: credentialSubjectID.String(),
+			CredentialType:      credentialType,
+		}
+		// Store credential's properties
+		keys, values := indexJSONObject(currCred.CredentialSubject[0].(map[string]interface{}), nil, nil, "credentialSubject")
+		for i, key := range keys {
+			if key == "credentialSubject.id" {
+				// present as column, don't index
+				continue
+			}
+			newCredential.Properties = append(newCredential.Properties, credentialPropertyRecord{
+				ID:    newCredential.ID,
+				Key:   key,
+				Value: values[i],
+			})
+		}
+		newPresentation.Credentials = append(newPresentation.Credentials, newCredential)
+	}
+	return &newPresentation, nil
+}
+
+func (s *sqlStore) get(serviceID string, startAt Timestamp) ([]vc.VerifiablePresentation, *Timestamp, error) {
+	var rows []presentationRecord
+	err := s.db.Order("timestamp ASC").Find(&rows, "service_id = ? AND timestamp > ?", serviceID, int(startAt)).Error
+	if err != nil {
+		return nil, nil, fmt.Errorf("query service '%s': %w", serviceID, err)
+	}
+	timestamp := startAt
+	presentations := make([]vc.VerifiablePresentation, 0, len(rows))
+	for _, row := range rows {
+		presentation, err := vc.ParseVerifiablePresentation(row.PresentationRaw)
+		if err != nil {
+			return nil, nil, fmt.Errorf("parse presentation '%s' of service '%s': %w", row.PresentationID, serviceID, err)
+		}
+		presentations = append(presentations, *presentation)
+		timestamp = Timestamp(row.Timestamp)
+	}
+	return presentations, &timestamp, nil
+}
+
+func (s *sqlStore) search(serviceID string, query map[string]string) ([]vc.VerifiablePresentation, error) {
+	propertyColumns := map[string]string{
+		"id":                   "cred.credential_id",
+		"issuer":               "cred.credential_issuer",
+		"type":                 "cred.credential_type",
+		"credentialSubject.id": "cred.credential_subject_id",
+	}
+
+	stmt := s.db.Model(&presentationRecord{}).
+		Where("service_id = ?", serviceID).
+		Joins("inner join discoveryservice_credentials cred ON cred.presentation_id = discoveryservice_presentations.id")
+	numProps := 0
+	for jsonPath, value := range query {
+		if value == "*" {
+			continue
+		}
+		// sort out wildcard mode
+		var eq = "="
+		if strings.HasPrefix(value, "*") {
```

Comment:
Prefix wildcards are dangerous. Some DBs require this to be enabled. You could also have a fancy search by inserting `.` wildcards between every character.

Reply:
> Prefix wildcards are dangerous. Some DBs require this to be enabled. 
Source? And I don't expect there to be millions of rows, so should be OK? Although I'm also fine with disabling prefix wildcard search.
> You could also have a fancy search by inserting . wildcards between every character.

?

Reply:
Actually using `%` instead of `.`. You type `zrg` as user. You expand it to `z%r%g%`. This will match `zorg`, `zerg` and `zaterdag` (and others). Benefit is that you can use a binary index on the first `z`. Depending on the letter this reduces search space 10 to 50 fold.

Reply:
Still don't understand the `.`?


I think we should disallow wildcards-in-the-middle, I believe it's not supported by at least some of them (MySQL). 
So just postfix (and maybe prefix)?